Feature/arrays

skas102 opened the pull request.

Commit:
Add Nodes and modify parser visitor

Commit:
Fix abstract visitor

Commit:
Implement ArrayType and simple tests

Commit:
Remove unnecessary nodes

Commit:
Partial implementation

Commit:
Fix Array creation

Commit:
Allow Nested ArrayLengthCreation

Commit:
Fix array creation

Commit:
Finalize Array initialization

Commit:
Fix Array Assignment

Commit:
Simplify array initialization syntax

Commit:
Merge branch 'master' into feature/arrays

Commit:
Handle array variable declaration

Commit:
Add additional tests

Commit:
Test MultipleAssignment for Arrays

tk-codes commented on checker/typecheck/type_check_visitor.go:
```diff
@@ -320,11 +320,17 @@ func (v *typeCheckVisitor) VisitStructFieldAssignmentNode(node *node.StructField
-// VisitTypeNode currently does nothing
-func (v *typeCheckVisitor) VisitTypeNode(node *node.TypeNode) {
+// VisitBasicTypeNode currently does nothing
+func (v *typeCheckVisitor) VisitBasicTypeNode(node *node.BasicTypeNode) {
```
It is not relevant to override VisitBasicTypeNode because there is no implementation for basic type node.
Please remove it.

tk-codes commented on checker/typecheck/type_check_visitor.go:
```diff
@@ -320,11 +320,17 @@ func (v *typeCheckVisitor) VisitStructFieldAssignmentNode(node *node.StructField
 }
 
+// VisitArrayTypeNode visits the type node
+func (v *typeCheckVisitor) VisitArrayTypeNode(node *node.ArrayTypeNode) {
```
This is also not relevant

Commit:
Remove ArrayNode as it's not required

tk-codes commented on parser/parser.go:
```diff
@@ -280,17 +290,23 @@ func (p *Parser) parseStatementWithFixToken() node.StatementNode {
-		AbstractNode: varType.AbstractNode,
-		Type:         varType,
-		Identifier:   id,
+		AbstractNode: node.AbstractNode{
```
Use `abstractNode` argument

tk-codes commented on parser/parser.go:
```diff
@@ -311,10 +327,12 @@ func (p *Parser) parseMultiVariableStatement(varType *node.TypeNode, id string)
-		Types:        types,
-		Identifiers:  ids,
-		FuncCall:     p.parseFuncCall(p.parseDesignator()),
+		AbstractNode: node.AbstractNode{
```
You could use `p.newAbstractNodeWithPos(varType.Pos())` helper function

tk-codes commented on parser/parser.go:
```diff
@@ -402,12 +420,47 @@ func (p *Parser) parseReturnStatement() *node.ReturnStatementNode {
+	return typeNode
+}
+
+func (p *Parser) parseType() node.TypeNode {
```
Function order convention: `parseType` comes before `parseTypeWithIdentifier`

tk-codes commented on parser/parser.go:
```diff
@@ -402,12 +420,47 @@ func (p *Parser) parseReturnStatement() *node.ReturnStatementNode {
+	}
+	p.check(token.CloseBracket)
+	arrayTypeNode := &node.ArrayTypeNode{
+		AbstractNode: node.AbstractNode{
```
Use helper function `newAbstractNodeWithPos`

tk-codes commented on parser/parser_expression.go:
```diff
@@ -216,6 +215,10 @@ func (p *Parser) parseDesignator() node.DesignatorNode {
 	return left
 }
 
+func (p *Parser) parseDesignator() node.DesignatorNode {
```
Move before parseDesignatorWithIdentifier

tk-codes commented on parser/parser_expression.go:
```diff
@@ -243,12 +246,85 @@ func (p *Parser) parseCreation() node.ExpressionNode {
+	return arrayInitialization
+}
+
+func (p *Parser) parseArrayCreation(abstractNode node.AbstractNode, identifier string) node.ExpressionNode {
```
Move before `parseArrayInitialization`

tk-codes commented on parser/parser_expression.go:
```diff
@@ -243,12 +246,85 @@ func (p *Parser) parseCreation() node.ExpressionNode {
+	arrayType := p.parseTypeWithIdentifier(abstractNode, identifier)
+
+	// Initialization using values
+	if valueCreationNode, ok := arrayType.(*node.ArrayTypeNode); ok {
```
You cast `arrayType` to `ArrayTypeNode`. Logically, it is not comprehensible at first why it has to be like that.

I would refactor it like that:
- first create a type
```
       typeNode := &node.BasicTypeNode{
		AbstractNode: abstractNode,
		Identifier:   identifier,
	}
```
- skip open bracket `[`
- Then create ArrayValueCreationNode if not close bracket

```
        if !p.isSymbol(token.CloseBracket) {
		// create ArrayValueCreationNode
	}
```

With that, you can remove the special case in `parseArrayType`.

skas102 replied:
I think this solution will not work with nested arrays

skas102 replied:
```
 if !p.isSymbol(token.CloseBracket) {
		// create ArrayValueCreationNode
	}
```

Theres a general mistake in there. ValueCreation is new int[]{1, 2}...

tk-codes replied:
@rpfister102 It works like this: https://github.com/bazo-blockchain/lazo/pull/64/commits/e8c898e99f7e1c7e3a005c83e896d0749b8186ae

tk-codes reviewed (changes requested):
The following cases doesn't work:
``` 
int[] x = new int[5]
int[] x = new int[]{1,2,3}

tk-codes committed:
Merge branch 'master' into feature/arrays

tk-codes committed:
Fix array variable initialization

Expression shouldn't check for newline. VariableStatement will check for newline

Commit:
Test parser default cases

Commit:
Fix concerns

tk-codes committed:
Fix ArrayValueCreation newline bug

tk-codes committed:
Simplify parseArrayCreation and parseArrayType

- Decouple parseArrayType from array creation
- Avoid scattered parse logic: Array creation logic knows when to call parseArrayType. parseArrayType shouldn't do a workaround for array value creation case.
